sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqIO.java

```diff
@@ -233,26 +255,73 @@ public Read withQueue(String queue) {
+     * used. Declaring either type of object ensures that one of that name and of the specified
+     * properties exists, creating it if necessary.
      *
-     * @param queueDeclare If {@code true}, {@link RabbitMqIO} will declare the queue. If another
-     *     application declare the queue, it's not required.
+     * <p>NOTE: When declaring a queue or exchange that already exists, the properties specified in
+     * the declaration must match those of the existing queue or exchange. That is, if you declare a
+     * queue to be non-durable but a durable queue already exists with the same name, the
+     * declaration will fail. When declaring a queue, RabbitMqIO will declare it to be non-durable.
+     *
+     * @param queueDeclare If {@code true}, {@link RabbitMqIO} will declare a non-durable queue. If
+     *     another application created the queue, this is not required and should be set to {@code
+     *     false}
      */
     public Read withQueueDeclare(boolean queueDeclare) {
       return builder().setQueueDeclare(queueDeclare).build();
     }
 
     /**
-     * Instead of consuming messages on a specific queue, you can consume message from a given
-     * exchange. Then you specify the exchange name, type and optionally routing key where you want
-     * to consume messages.
+     * In AMQP, messages are published to an exchange and routed to queues based on the exchange
+     * type and a queue binding. Most exchange types utilize the routingKey to determine which
+     * queues to deliver messages to. It is incumbent upon the developer to understand the paradigm
+     * in place to determine whether to declare a queue, what the appropriate binding should be, and
+     * what routingKey will be in use.
+     *
+     * <p>This function should be used if the Beam pipeline will be responsible for declaring the
+     * exchange. As a result of calling this function, {@code exchangeDeclare} will be set to {@code
+     * true} and the resulting exchange will be non-durable and of the supplied type. If an exchange
+     * with the given name already exists but is durable or is of another type, exchange declaration
+     * will fail.
+     *
+     * <p>To use an exchange without declaring it, especially for cases when the exchange is shared
+     * with other applications or already exists, use {@link #withExchange(String, String)} instead.
+     *
+     * @see
+     *     "https://www.cloudamqp.com/blog/2015-09-03-part4-rabbitmq-for-beginners-exchanges-routing-keys-bindings.html"
+     *     for a write-up on exchange types and routing semantics
      */
-    public Read withExchange(String name, String type, String routingKey) {
-      checkArgument(name != null, "name can not be null");
-      checkArgument(type != null, "type can not be null");
-      return builder().setExchange(name).setExchangeType(type).setRoutingKey(routingKey).build();
+    public Read withExchange(String name, String type, @Nullable String routingKey) {
```

Comment:
I think it might make sense to rename this function and the other one respectively to `withDeclareExchange` and `withExistingExchange`. Though, keep the original functions for backwards compatibility (but mark them deprecated, make them simply delegate to the non-deprecated versions, and make their documentation a one-liner directing to those versions too)

Same applies to the write transform.

Reply:
Not a bad idea. I can do this in a follow-up PR (this one has been merged).